service/tablet_allocator.cc

```diff
@@ -715,6 +717,99 @@ class load_balancer {
         return (it != _table_load_stats->tables.end()) ? &it->second : nullptr;
     }
 
+    future<migration_plan> make_merge_colocation_plan(node_load_map& nodes) {
+        migration_plan plan;
+        table_resize_plan resize_plan;
+
+        for (auto&& [table, tmap_] : _tm->tablets().all_tables()) {
+            auto& tmap = *tmap_;
+            if (!tmap.needs_merge()) {
+                continue;
+            }
+
+            // Also filter out replicas that don't belong to the DC being worked on.
+            auto sorted_new_replicas = [this, &nodes] (const tablet_desc& t) {
+                auto ret = sorted_replicas_for_tablet_load(*t.info, t.transition);
+                std::ranges::remove_if(ret, [&] (tablet_replica r) { return !nodes.contains(r.host); });
+                return ret;
+            };
+
+            auto migrating = [] (const tablet_desc& t) {
+                return bool(t.transition);
+            };
+
+            auto first_non_matching_index = [] (tablet_replica_set r1, tablet_replica_set r2) -> std::optional<unsigned> {
+                assert(r1.size() == r2.size());
+                for (size_t i = 0; i < r1.size(); i++) {
+                    if (r1[i] != r2[i]) {
+                        return i;
+                    }
+                }
+                return std::nullopt;
+            };
+
+            auto create_migration_info = [] (global_tablet_id gid, tablet_replica src, tablet_replica dst) {
+                auto kind = (src.host != dst.host) ? tablet_transition_kind::migration : tablet_transition_kind::intranode_migration;
+                return tablet_migration_info{kind, gid, src, dst};
+            };
+
+            bool all_colocated = true;
+            co_await tmap.for_each_sibling_tablets([&] (tablet_desc t1, std::optional<tablet_desc> t2_opt) -> future<> {
+                // Be optimistic about migrating tablets, as if they succeeded.
+                // Merge finalization will have to recheck that all sibling tablets are co-located.
+                // Not maintaining replica order, since with RF=N, we cannot swap replicas
+
+                if (!t2_opt) {
+                    on_internal_error(lblogger, format("Unable to find sibling tablet during co-location, with tablet count {}, for table {}",
+                                                       tmap.tablet_count(), table));
+                }
```

Comment:
May as well not iterate at all for single tablets.

Reply:
This works as an assertion, since balancer should never emit a merge decision for a single-tablet table. Also that tablet count must be power-of-two. It's an invariant today, but I thought it doesn't hurt to check, works as a documentation too. In practice, we shouldn't bump into this unless we have a terrible bug.